packages/axelar-soroban-std-derive/src/lib.rs

```diff
@@ -220,3 +222,119 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
         }
     }.into()
 }
+
+/// Implements the Event and EventTestUtils traits for a Soroban contract event.
+///
+/// Fields without a `#[data]` attribute are used as topics, while fields with `#[data]` are used as event data.
+/// The event name can be specified with `#[event_name(...)]` or will default to the struct name in snake_case (minus "Event" suffix).
+///
+/// # Example
+/// ```rust
+/// # use soroban_sdk::{String, Address};
+/// # use axelar_soroban_std_derive::IntoEvent;
+/// #[derive(Debug, PartialEq, IntoEvent)]
+/// pub struct TransferEvent { // event name is "transfer"
+///     pub from: Address,
+///     pub to: Address,
+///     #[data]
+///     pub amount: String,
+/// }
+/// ```
+#[proc_macro_derive(IntoEvent, attributes(data, event_name))]
+pub fn derive_into_event(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let event_impl = derive_event_impl(&input);
+    let test_impl = derive_test_impl(&input);
+
+    quote! {
+        #event_impl
+        #test_impl
+    }
+    .into()
+}
+
+fn derive_event_impl(input: &DeriveInput) -> proc_macro2::TokenStream {
+    let name = &input.ident;
+    let event_name = get_event_name(input);
+    let (topic_fields, data_fields) = get_fields(input);
+
+    let topic_idents = topic_fields.iter().map(|(ident, _)| ident);
+    let data_idents = data_fields.iter().map(|(ident, _)| ident);
+
+    quote! {
+        impl axelar_soroban_std::events::Event for #name {
+            fn topics(&self, env: &soroban_sdk::Env) -> impl soroban_sdk::Topics + core::fmt::Debug {
+                (
+                    soroban_sdk::Symbol::new(env, #event_name),
+                    #(soroban_sdk::IntoVal::<soroban_sdk::Env, soroban_sdk::Val>::into_val(&self.#topic_idents, env),)*
+                )
+            }
+
+            fn data(&self, env: &soroban_sdk::Env) -> impl soroban_sdk::IntoVal<soroban_sdk::Env, soroban_sdk::Val> + core::fmt::Debug {
+                let mut vec = soroban_sdk::Vec::<soroban_sdk::Val>::new(env);
+                #(
+                    vec.push_back(soroban_sdk::IntoVal::<soroban_sdk::Env, soroban_sdk::Val>::into_val(&self.#data_idents, env));
+                )*
```

Comment:
instead of duplicating push_back, use the vec macro to do something like the following?
```suggestion
                soroban_sdk::vec![env, #(soroban_sdk::IntoVal::<_, soroban_sdk::Val>::into_val(&self.#data_idents, env))*]
```